src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService.cs

```diff
@@ -0,0 +1,124 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.FindUsages;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.SymbolMapping;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.InheritanceMargin
+{
+    using SymbolKeyAndLineNumberArray = ImmutableArray<(SymbolKey, int lineNumber)>;
+
+    internal abstract partial class AbstractInheritanceMarginService : IInheritanceMarginService
+    {
+        /// <summary>
+        /// Given the syntax nodes to search,
+        /// get all the method, event, property and type declaration syntax nodes.
+        /// </summary>
+        protected abstract ImmutableArray<SyntaxNode> GetMembers(IEnumerable<SyntaxNode> nodesToSearch);
+
+        /// <summary>
+        /// Get the token that represents declaration node.
+        /// e.g. Identifier for method/property/event and this keyword for indexer.
+        /// </summary>
+        protected abstract SyntaxToken GetDeclarationToken(SyntaxNode declarationNode);
+
+        protected abstract string GlobalImportsTitle { get; }
+
+        public async ValueTask<ImmutableArray<InheritanceMarginItem>> GetInheritanceMemberItemsAsync(
+            Document document,
+            TextSpan spanToSearch,
+            CancellationToken cancellationToken)
```

Comment:
this was a move, but the logic here did change somewhat.  namely that we can still call into OOP, even if we have no membersymbolkeys (since we may have global imports).